### ISSUE TITLE
"WARNING: The windows target is oriented towards MSVC; maybe you want --os=cygwin or --os=mingw"

### ISSUE THREAD

**Issue by oviano** [randombit/botan#3703]
So what does this mean in the context of building with the clang-cl compiler optionally installed with MSVC?

The only side-effect I have noticed so far since switching from the MSVC cl.exe compiler to clang-cl.exe was the need to explicitly link the executable with clang_rt.builtins-x86_64.lib so perhaps the warning above can be ignored in this context?

Or am I supposed to be telling Botan that the CC is still msvc, when taking this approach?

**Comment by randombit** [comment]
Interesting question! I think this warning is now obsolete; we previously would do some feature enablement based on the target being Windows vs Cygwin, but that's been all removed now I think.

Probably for `clang-cl` the correct target is `windows`.

I'll look into if we can create a CI target for this to make sure it works reliably.

**Comment by oviano** [comment]
Yep. I'm using the target Windows, and the CMAKE_COMPILER_ID is coming through as clang.

The only other change I had/wanted to make was that using clang-cl.exe brings in a warning in the code below, so you will see I expanded the #if defined condition to also include BOTAN_BUILD_COMPILER_IS_CLANG.

Otherwise it moans about deprecated/insecure getenv() function.

```
bool OS::read_env_variable(std::string& value_out, std::string_view name_view) {
   value_out = "";

   if(running_in_privileged_state())
      return false;

#if defined(BOTAN_TARGET_OS_HAS_WIN32) && (defined(BOTAN_BUILD_COMPILER_IS_MSVC) || defined(BOTAN_BUILD_COMPILER_IS_CLANG))
   const std::string name(name_view);
   char val[128] = {0};
   size_t req_size = 0;
   if(getenv_s(&req_size, val, sizeof(val), name.c_str()) == 0) {
      value_out = std::string(val, req_size);
      return true;
   }
#else
   const std::string name(name_view);
   if(const char* val = std::getenv(name.c_str())) {
      value_out = val;
      return true;
   }
#endif
```

**Comment by Zimbrando** [comment]
I'm trying to compile Botan version 2.19.5 using Clang-CL on Windows, as part of setting up a project for [QSsh](https://github.com/sandsmark/QSsh). However, I'm running into multiple errors during the build process.

I first tried configuring the build using the following command:

`python3 configure.py --cc msvc --cc-bin="C:\PROGRA~1\LLVM\bin\clang-cl.exe"  --os windows --with-debug-info --cxxflags="/EHsc"`

This leads to multiple errors like this one:
```
build\include\botan/internal/simd_avx2.h(167,36): error: cannot initialize a parameter of type '__m128i' (vector of 2 'long long' values) with an lvalue of type '__m256i' (vector of 4 'long long' values)                                           m_avx2 = _mm256_add_epi32(m_avx2, other.m_avx2); 
```

To narrow down the issue, I tried disabling certain modules, but then I encountered another error related to rdrand (as mentioned by @oviano).

I'm not very familiar with compilers and build configurations, so I'm not sure where to look or how to resolve these issues. Could anyone point me in the right direction or suggest potential fixes? Any help would be greatly appreciated!

Windows 10, LLVM 10.0.0
